Code base logging Blog Post : issue https://github.com/mlflow/mlflow-website/issues/87 (#95)
mlflow/mlflow-website: website/blog/2024-09-01-model-as-code-logging/index.md

---
title: Model-as-Code Logging in MLflow for Enhanced Model Management
tags: [pyfunc, genai]
slug: model_as_code

> authors: [awadelrahman-ahmed, daniel-liden]

Author: Thanks for the excellent guidance!

thumbnail: img/blog/thumbnail_model_as_code.gif
---

> We all—well, most of us—remember November 2022 when the public release of ChatGPT by OpenAI marked a significant turning point in the world of AI. While GenAI had been evolving for some time, ChatGPT, built on OpenAI's GPT-3.5 architecture, quickly captured the public’s imagination. This led to an explosion of interest in GenAI, both within the tech industry and among the general public.

Reviewer: The hyphen conversational abrupt stall threw me for a loop (I thought it was a typo). Perhaps encapsulate in parenthesis?

> On the tools side, MLflow continues to solidify its position as the favorite tool for MLOps among the ML community. However, the rise of GenAI has introduced new needs in how we use MLflow. One of these new challenges is how we log model artifacts in MLflow. If you’ve used MLflow before (and we bet you have), you’re probably familiar with the `mlflow.log_model()` function and how it efficiently **pickles** model artifacts (I’ve bolded "pickles" because it’s a key term in this post).

Reviewer: It might be worthwhile to convert the `pickle` reference to a link to `cloudpickle` https://github.com/cloudpipe/cloudpickle

> Particularly with GenAI, there’s a new requirement: logging the model "as code," instead of serializing it into a pickle file. And guess what? This need isn’t limited to GenAI models. So, in this post we will explore this concept and how MLflow has adapted to meet this new requirment.

Reviewer: Can we convert the reference of `models from code` to a link to https://mlflow.org/docs/latest/models.html#models-from-code ?

Author: please check if this is what you meant!

> You will notice that this feature is implemented at a very abstract level, allowing you to log any model "as code", whether it’s GenAI or not! We like to think of it as a generic approach, with GenAI models being just one of its use cases. So, in this post, we’ll explore this new approach, "Model-as-Code logging".

Reviewer: We've renamed the feature to be "models from code" - can you update the references?

Author: you mean chagning the term "Model-as-Code" loging to "Model-from-Code" ,  "models from code" or  "models-from-code" ?

Reviewer: The feature reference can be referred to as either "Models from Code" or "models from code". We don't need the hyphens.

By the end of this post you should be able to answer the three main questions: "what", "why" and "how" to do Model-as-Code logging.

> ## What Is Actually Model-as-Code Logging?

Reviewer 2: ```suggestion
## What Is Model-as-Code Logging?
```

In fact, when MLflow announced this feature, it got us thinking in a more abstract way about the concept of a "model"! You might find it interesting, too, if you zoom out and consider a model as a mathematical representation or function that describes the relationship between input and output variables. At this level of abstraction, a model can be many things!

> At this level, one might recognize that a model, as an object or artifact, represents just one form of what a model can be, even if it’s the most popular in the ML community. But if you think about it, a model can also be a simple mapping function — just a piece of code — or even code that sends API requests to another service that doesn’t necessarily reside within your "premises" (e.g., OpenAI APIs).
> 

Reviewer 2: ```suggestion
At this level, one might recognize that a model, as an object or artifact, represents just one form of what a model can be, even if it’s the most popular in the ML community. But if you think about it, a model can also be a simple mapping function—just a piece of code—or even code that sends API requests to another service that doesn’t necessarily reside within your "premises" (e.g., OpenAI APIs).

```

We'll explain the detailed workflow of how to log Mode-as-Code later in the post, but for now, let's consider it at a high level with two main steps: first, writing your model code, and second, logging your model as code. This will look like the folowing figure:

![High Level Model-as-Code Logging Workflow](model_as_code1.png)

> 🔴 It's important to note that when we refer to "model code," we're talking about code that can be treated as a model itself. This means it's **not** your training code that generates a trained model object, but rather the step-by-step code that is executed as a model itself.

Reviewer 2: Maybe "inference code"?

Author: No, it is the "training code" is what the reader might confuse with the "model code" that we are talking about! ML engineers used to write code to produce model (objects) (e.g., Sklean linear regression) and they might think this is what this feature is logging as code, which is not correct!

Reviewer 2: I meant "inference code" instead of or in addition to "model code," not instead of "training code" — reflecting the workflow of separating training and inference code.

Clear enough without the change, regardless.

The Model-as-Artifact logging can be broken down into three high-level steps as in the following figure: first, creating the model as an object (whether by training it or acquiring it), second, serializing it (usually with Pickle or a similar tool), and third, logging it as an object.

![High Level Model-as-Artifact Logging Workflow](model_as_code2.png)

💡 So, the main distinction between the popular Model-as-Artifact logging and Model-as-Code logging is that in the former, we log the model object itself—whether it's a model you've trained or a pre-trained model you've acquired. In the latter, however, we log the code that _represents_ your model.

> ## When Do You Need Model-as-Code Logging?

Author: Please check if this is what you meant

For our Model-as-Code logging, we’ll particularly use the `PythonModel` class within `pyfunc`. This class in the MLflow Python client library allows us to create and manage Python functions as MLflow models. It enables us to define a custom function that processes input data and returns predictions or results. This model can then be deployed, tracked, and shared using MLflow's features.

It seems to be exactly what we're looking for—we have some code that serves as our model, and we want to log it! That's why you'll soon see `mlflow.pyfunc.PythonModel` in our code example!

Now, each time we need to implement Model-as-Code logging, we create _Two_ separate Python files:

1.  The first contains our model code (let's call it `model_code.py`). This file contains a class that inherits from the `mlflow.pyfunc.PythonModel` class.
    It contains your model logic. It could be your calls to OpenAI APIs, CLV (Customer Lifetime Value) model, or your causal discovery code. We'll use a very simple example in our 101 example.

    📌 But wait! IMPORTANT:

>                 - Yout `model_code.py` needs to call (i,e; include) `mlflow.models.set_model()` to set the model, which is crucial for loading the model back using `load_model` for inference. You will notice this in the example.

Reviewer: ```suggestion
                - Your `model_code.py` script needs to call (i.e., include) `mlflow.models.set_model()` to set the model, which is crucial for loading the model back using `load_model` for inference. You will notice this in the example.
```

We could propose a novel machine learning model to **predict** the area of a circle based on its diameter, rather than relying on the traditional formula (because nothing called π (pi) has been discovered).

While this idea might seem humorous, it serves as a good example of how abstract the concept of a model can be. If you agree that **calculating** the area of a circle without the discovery of π (pi) could be treated as a **prediction** problem, then we’ve reached a point of understanding the notion of MLflow treating models as code.

Now, we can log that calculation as code and refer to it as a prediction. Sounds natural, doesn’t it?

So now let us write our Two files:

#### 1. Our `model_code.py` file :

```python
import mlflow
import math

class CircleAreaModel(mlflow.pyfunc.PythonModel):
    def predict(self, context, model_input, params=None):
        return [math.pi * (r ** 2) for r in model_input]

# It's important to call set_model() so it can be loaded for inference
# Also, note that it is set to an instance of the class, not the class itself.
mlflow.models.set_model(model=CircleAreaModel())
```

#### 2. Our `driver.py` file :

> This can be a notebook as well. Here is it essential contents:

Reviewer: ```suggestion
This can be defined within a notebook as well. Here are its essential contents:
```
